src/domain/customscript-identity.ts

```diff
@@ -374,53 +373,59 @@ export class CustomScriptIdentity
               }
             }
 
-            const leafScript =
-              input.tapLeafScript && input.tapLeafScript.length > 0
-                ? input.tapLeafScript[0].script.toString('hex')
-                : this.getFirstAutoSpendableTapscriptPath(cachedAddrInfos);
+            let leafScript = undefined;
+            if (input.tapLeafScript && input.tapLeafScript.length > 0) {
+              leafScript = input.tapLeafScript[0].script.toString('hex');
+            } else {
+              leafScript = this.getFirstAutoSpendableTapscriptPath(cachedAddrInfos);
+              cachedAddrInfos.contract.getTaprootTree()
+              // if we use the auto-spendable leaf we need to add the tapLeafScript to the input
+              if (leafScript) {
+                if (!cachedAddrInfos.result.taprootInternalKey) {
+                  throw new Error('marina fails to sign input (no taproot internal key)');
+                }
+
+                const tree = cachedAddrInfos.contract.getTaprootTree()
+                const leaf = { scriptHex: leafScript }
+                const leafHash = bip341.tapLeafHash(leaf)
+
+                // witnesses func will throw if the leaf is not a valid leaf
+                const taprootSignScriptStack = bip341
+                  .BIP341Factory(this.ecclib)
+                  .taprootSignScriptStack(
+                    Buffer.from(cachedAddrInfos.result.taprootInternalKey, 'hex'),
+                    { scriptHex: leafScript },
+                    tree.hash,
+                    bip341.findScriptPath(tree, leafHash)
+                  );
+
+                pset.data.inputs[index].tapLeafScript = pset.data.inputs[index].tapLeafScript?.slice(1); // clear tapLeafScript first (we'll overwrite it)
+                pset.updateInput(index, {
+                  tapLeafScript: [
+                    {
+                      leafVersion: 0xc4, // elements tapscript version
+                      script: Buffer.from(leafScript, 'hex'),
+                      controlBlock: taprootSignScriptStack[1],
```

Comment:
not really understand this code: ideally user already updated input before signing?

Reply:
This is in case there is no leaf on the input (and it's not key path). In that case marina will try to add the "auto-spendable" leaf if any exist. we need it for the finalize step.